Prospress/action-scheduler
# Throw an exception (mark as failed) scheduled actions without any callbacks

This change is geared toward handling of scheduled actions that do not have any callbacks and, more broadly, recurring actions that continually and consistently fail.

Example: imagine a plugin has scheduled an hourly recurring action, but that same plugin is subsequently deleted without any opportunity to perform clean-up (or, it is deactivated normally but does not contain uninstall logic). Currently, those recurring actions will continue to fire and:

- Each time the action status will be updated to `complete` (so it will seem like they are running successfully).
- A replacement action will be scheduled.

Therefore, we can end up with a category of repeating zombie actions, clogging up the queue unnecessarily. This change attempts to address that problem and closes #805.

### How it works

Just because an action fails once doesn't mean it will always fail: so, we need to bake in some amount of tolerance. The way I propose we handle this is two-fold:

- When an action does not have any callbacks, mark it as `failed`.
- If it is a recurring action, check to see if it is *consistently* failing and, if so, don't reschedule it.
    - By default, the criteria for *"consistent failure"* is when the most recent 5 actions with the same hook, argument set, and schedule have all failed.
    - This will be adjustable via a filter.

### Testing

To make testing easier, I suggest temporarily stopping Action Scheduler from being triggered automatically:

- Add `define( 'DISABLE_WP_CRON', true );` to your `wp-config.php` file.
- Add `add_filter( 'action_scheduler_allow_async_request_runner', '__return_false' );` to a suitable location, such as a mu-plugin file.
- If you have any crontab entries that provide alternative ways of running cron or the Action Scheduler queue, suspend them.
- Also, to give yourself a 'clean view' of things, you may also wish to delete all existing actions: `wp db query "DELETE FROM $(wp db prefix)actionscheduler_actions"`

Now, processing of the queue should be entirely manual. Next:

- Create a new recurring action without a callback: `wp eval "as_schedule_recurring_action( time() - 1, 2, 'no_callback' );"`
- Visit **Tools ▸ Scheduled Actions** and locate action `no_callback` (should be pending), then run it manually.
- It should fail, and a replacement should be scheduled.
- Rinse and repeat a few times.
- In the end, you should find you have 5 failed scheduled actions for the hook `no_callback` and no more pending actions of that name should have been scheduled.

### Notes

- This is one facet of an effort to address zombie actions; we need to follow-up with further work that prevents unintentional duplication of actions (via race conditions).
- You will notice a large number of updates to tests that seem to be unrelated: that's because those tests relied on the previous behavior of Action Scheduler, where actions would be marked complete even if they didn't have a callback. Most of these updates are isolated inside commits https://github.com/woocommerce/action-scheduler/pull/806/commits/8276036dd57d3ba8ec8249e55a386c5acb1bfc4c and https://github.com/woocommerce/action-scheduler/pull/806/commits/37b8a1efb8911efc657609c0aa11c3392e3a4b86. The assertions specific to the core change can be [found here](https://github.com/woocommerce/action-scheduler/pull/806/files#diff-2fd59945ebe22bbea56dd19195fec979d5ee027dc705ac6b6af4eb01defad288).

### Changelog

> Enhancement - If a recurring action is found to be consistently failing, it will stop being rescheduled.
